examples/experimental/README.md

```diff
@@ -0,0 +1,70 @@
+- It is not clear which methods a new protocol needs to override to conform to the public TFE API (this API is largely implicitly defined and not stated explicitly anywhere)
+- BackingTensors being wrapped up in factories are impossible to test
+- Excessive code duplication: each operation is implemented at least once for each BackingTensor and each protocol
+
+
+
+### Overview
+The guiding philosophy of the refactor will be to replace TFE's major features with native TensorFlow abstractions that are covered by the TF Stability Guarantees for the lifetime of TF 2.0.  The goals of doing so include:
+- Upgrading TF Encrypted to TensorFlow 2.0
+- Leverage as much of the existing TensorFlow ecosystem with minimal rewriting and/or maintenance
+- Continue to rely on the stability guarantee of the TensorFlow Public API
+- Improve our own code modularity, extensibility, and sustainability and reduce code duplication
+
+Concretely, we will replace our own type system of custom tensors with a light set of hooks over existing tensor types in TensorFlow.  The current Protocol class would be replaced by custom subclasses of `tf.distribute.Strategy` and `tf.distribute.StrategyExtended` that define the necessary protocol operations and handle any distributed computation that's required by them.
+
+By doing so, we hope to satisfy each of the above goals, as well as all potential future requirements detailed by @mortendahl (see below)
+
+### Details
+- `Protocol` class replaced by custom implementations of `tf.distribute.Strategy` and `tf.distribute.StrategyExtended`
+  - The Strategy abstraction functions as our entry-point into the rest of the higher-level TF APIs
+  - The current strategy scope will determine which protocol kernel is used behind the scenes when calling a tfe op, e.g. `tfe.matmul`, similar to how it works now with Protocols
+  - For n-party protocols, this API provides a convenient set of tools for orchestrating distributed computation in a way that can induce out-of-the-box support for both tf.keras and the Estimator API
+  - Note these need not operate in the multi-party setting: see [`tf.distribute.OneDeviceStrategy`](https://www.tensorflow.org/api_docs/python/tf/distribute/OneDeviceStrategy) for an example of a single device strategy
+- Replace `BackingTensor` (`BackingVariable`) with `tf.Tensor` (`tf.Variable`)
+  - Allows us to use `Strategy` instead of `Protocol` (since Strategies must run on native TF Tensors/Variables)
+  - Allows us to use the `tf.custom_gradient` strategy to [bring back autograd](https://github.com/tf-encrypted/tf-encrypted/issues/244)
+  - Custom `BackingTensor` types can be implemented through `tf.VariantTensor`, similar to how the [`tf-big`](https://github.com/tf-encrypted/tf-big) project uses libgmp behind the scenes.  It may also be possible to use Tensors from another Graph to source the VariantTensor, which would allow us to reproduce our existing CRTTensor as well as the NativeTensor with an explicit modulus
```

Comment:
For TF Big we actually ended up with a python wrapper, ie a custom tensor type, since it's not very convenient to only deal with tensors of type `tf.variant` (in the case of TF Big). One problem is that you'd want to overload operations to allow `x + y`, yet this won't work on `tf.variant` tensors; to actually do an addition you have to call a custom op: `ops.add(x, y)`.

This may not be the end of the story though.

Reply:
Having the backing tensors as variant tensors and custom ops makes sense though, as there are likely some performance improvements to be had that way (both graph size but also evaluation runtime).

Reply:
yes, the other hope here is that by using a native TF tensor type we can try to bypass internal checks inside TF that have previously blocked us from hooking into the higher-level stuff (keras, autograd, etc.)